### DISCUSSION TITLE
cann't get event

### DISCUSSION THREAD

**Discussion by IamRezaMousavi** [tauri-apps/tauri#7675] [Q&A]
Hello guys
I wanted to send data from Rust to React, but React cannot receive it
Any ideas?

Rust Code:
```rust
// called in app.setup
pub fn main_window(app: AppHandle, show: bool) -> Window {
    let window = tauri::WindowBuilder::new(
        &app,
        "main",
        tauri::WindowUrl::App("index.html".parse().unwrap()),
    )
    .title("Germez")
    .visible(show)
    .build()
    .unwrap();

    let db = database::DataBase::new();
    let temp_db = database::TempDB::new();
    temp_db.create_table();
    db.set_db_table_to_default_value();
    let category = db.categories_list();
    window.emit("category-list", category).unwrap();

    let download_table_dict = db.return_items_in_download_table(None);
    window
        .emit("download-table-dict", download_table_dict)
        .unwrap();

    window
}
```

Tsx Code:
```tsx
useEffect(() => {
    const unlisten = listen('download-table-dict', (event) => {
      setMessage(`message: ${event.payload}`);
    });
    
    return () => {
      unlisten.then(f => f());
    };
  }, []);
```

**Comment by IamRezaMousavi** [comment]
But React can receive event emitted from system-tray after clicked

**Comment by FabianLars** [comment]
This _should_ be a timing issue. If you emit events from inside the setup hook without having some sort of non-blocking delay (think of spawning a thread in the setup hook that does some work for a few seconds and _then_ emit the event), your javascript won't have had the time to completely run (or started at all) so the events are just lost.

**Reply by IamRezaMousavi** [reply]
How can I check javascript is completely runned or not?
Should I send a signal from javascript or is there a method for this?

**Reply by FabianLars** [reply]
> Should I send a signal from javascript

yeah, probably. In a few of my apps i refactored that into an `init` [tauri command](https://tauri.app/v1/guides/features/command/) because that feels more natural to me than the event variant.

> is there a method for this?

Nope, we'd love that too but there's just no reliable way to know when the js is done, especially with how js centric frontends are nowadays
